CoreScriptsRoot/Modules/VR/LaserPointer.lua

```diff
@@ -0,0 +1,548 @@
+--LaserPointer.lua
+--Implements the visual part of the VR laser pointer and implements
+--VR teleporting and movement mechanics
+--Written by Kyle, September 2016
+local CoreGui = game.CoreGui
+local RobloxGui = CoreGui:WaitForChild("RobloxGui")
+local ContextActionService = game:GetService("ContextActionService")
+local Players = game:GetService("Players")
+local SoundService = game:GetService("SoundService")
+local PathfindingService = game:GetService("PathfindingService")
+local Utility = require(RobloxGui.Modules.Settings.Utility) --todo: use common utility module when it's done
+
+local LocalPlayer = Players.LocalPlayer
+
+--Pathfinding sort of works, but it's very slow and does not handle slopes very well.
+--Use at your own risk.
+local usePathfinding = false
+
+local gamma, invGamma = 2.2, 1/2.2
+local function fromLinearRGB(color)
+	return Color3.new(color.r ^ gamma, color.g ^ gamma, color.b ^ gamma)
+end
+local function toLinearRGB(color)
+	return Color3.new(color.r ^ invGamma, color.g ^ invGamma, color.b ^ invGamma)
+end
+
+local CONTAINER_FOLDER_NAME = "VRCorePanelParts"
+local function setPartInGame(part, inGame)
+	if not part then
+		return
+	end
+	if inGame and not part:IsDescendantOf(game) then
+		local container = workspace.CurrentCamera:FindFirstChild(CONTAINER_FOLDER_NAME)
+		if not container then
+			coroutine.wrap(function()
+				part.Parent = workspace.CurrentCamera:WaitForChild(CONTAINER_FOLDER_NAME)
+			end)()
+		else
+			part.Parent = container
+		end
+	else
+		part.Parent = nil
+	end
+end
+
+--Teleport visual configuration
+local TELEPORT = {
+	ARC_COLOR_GOOD = fromLinearRGB(Color3.fromRGB(0, 162, 255)),
+	ARC_COLOR_BAD = fromLinearRGB(Color3.fromRGB(253, 68, 72)),
+	ARC_THICKNESS = 0.05,
+
+	PLOP_GOOD = "rbxasset://textures/ui/VR/VR Pointer Disc Blue.png",
+	PLOP_BAD = "rbxasset://textures/ui/VR/VR Pointer Disc Red.png",
+	PLOP_BALL_COLOR_GOOD = BrickColor.new("Bright green"),
+	PLOP_BALL_COLOR_BAD = BrickColor.new("Bright red"),
+	PLOP_BALL_SIZE = 0.5,
+	PLOP_SIZE = 2,
+	PLOP_PULSE_MIN_SIZE = 0,
+	PLOP_PULSE_MAX_SIZE = 2,
+
+	MAX_VALID_DISTANCE = 32,
+
+	BUTTON_DOWN_THRESHOLD = 0.95,
+	BUTTON_UP_THRESHOLD = 0.5,
+
+	MIN_VELOCITY = 10,
+	RANGE_T_EXP = 2,
+	G = 50,
+
+	PULSE_DURATION = 0.8,
+	PULSE_PERIOD = 1,
+	PULSE_EXP = 2,
+	PULSE_SIZE_0 = 0.25,
+	PULSE_SIZE_1 = 2,
+
+	BALL_WAVE_PERIOD = 2,
+	BALL_WAVE_AMPLITUDE = 0.5,
+	BALL_WAVE_START = 0.25,
+	BALL_WAVE_EXP = 0.8,
+
+	FLOOR_OFFSET = 4.5,
+
+	FADE_OUT_DURATION = 0.125,
+	FADE_IN_DURATION = 0.125,
+
+	CLEAR_AABB_SIZE = Vector3.new(2.5, 4, 2.5),
+
+	SUCCESS_SOUND = "rbxassetid://147722227",
+	FAIL_SOUND = "rbxassetid://138087015",
```

Comment:
I doubt we actually want these. I just forgot to remove them. The fail sound is the horn from The Price is Right ;)

That being said, need to talk to product/art about getting actual sounds, because audible confirmation is really nice.